### ISSUE TITLE
play-redis fails to round-trip Option through the cache due to serialization manifest/additional bindings in default Akka conf

### ISSUE THREAD

**Issue by dominics** [KarelCemus/play-redis#191]
## Setup

- play-redis version: 2.3.0
- Play version: 2.6.17
- Scala version: 2.12
- Akka version: 2.5.17

Default Akka configuration. Well, I'm actually using kryo serialization, but I only have a serialization binding for `java.io.Serializable`, so I don't think that's relevant. For other types, Akka should be using its own default serialization, i.e.:

```
    "akka" : {
        "actor" : {
[...]
            # Additional serialization-bindings that are replacing Java serialization are
            # defined in this section for backwards compatibility reasons. They are included
            # by default but can be excluded for backwards compatibility with Akka 2.4.x.
            # They can be disabled with enable-additional-serialization-bindings=off.
            # adds the protobuf serialization of pub sub messages to groups
            "additional-serialization-bindings" : {
[...]
                # reference.conf @ jar:file:~/.ivy2/cache/com.typesafe.akka/akka-remote_2.12/jars/akka-remote_2.12-2.5.17.jar!/reference.conf: 72
                "scala.None$" : "akka-misc",
                # reference.conf @ jar:file:~/.ivy2/cache/com.typesafe.akka/akka-remote_2.12/jars/akka-remote_2.12-2.5.17.jar!/reference.conf: 71
                "scala.Some" : "akka-misc"
            },
```

## Problem

In a test, I'm trying to round-trip an Option through the cache, like:

```scala
val cache: AsyncCacheApi = ???
(
  for {
    _ <- cache.remove("foo")
    _ <- cache.set("foo", Some("thing"))
    get <- cache.get[Option[String]]("foo")
  } yield get
).await.get must beSome("thing")
```

The value makes it to Redis, serialized:

```
1542075094.980718 [0 172.23.0.1:32914] "DEL" "foo"
1542075094.998764 [0 172.23.0.1:32914] "SET" "foo" "CgkKBXRoaW5nEBQ="
1542075095.005598 [0 172.23.0.1:32914] "GET" "foo"
```

But the test ultimately fails on the `Option.get`, because the value fails to be deserialized. Tracing this through to find the deserialization exception reveals Akka is using its [`MiscMessageSerializer`](https://github.com/akka/akka/blob/v2.5.17/akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala) (as shown in the above config) to serialize/deserialize `Option`s, and this is failing with:

> java.io.NotSerializableException: Unimplemented deserialization of message with manifest [scala.Some] in [akka.remote.serialization.MiscMessageSerializer]

### Deserialization manifests

Under the hood, play-redis seems to use [this method](https://github.com/akka/akka/blob/b4f799a7a9096f6bd1ef57335afa903799fa3492/akka-actor/src/main/scala/akka/serialization/Serialization.scala#L282) (`def deserialize[T](bytes: Array[Byte], clazz: Class[T]): Try[T]`) to deserialize values using Akka's serialization. This doesn't seem to work, because the manifest passed to `MiscMessageSerializer` ends up being the class name (`scala.Some`), whereas `MiscMessageSerializer` actually expects the string `C` in the case of an `Option` (as seen [here](https://github.com/akka/akka/blob/v2.5.17/akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala#L283)).

Note that you can reproduce this error by using the default Akka configuration, in combination with the `deserialize[T](...)` method:

```scala
val system: ActorSystem = ???
val serialization = SerializationExtension(system)
val serialized: Array[Byte] = serialization.serialize(Some("thing")).get
val deserialized: Some[String] = serialization.deserialize[Some[String]](serialized, classOf[Some[String]]).get
// fails with java.io.NotSerializableException: Unimplemented deserialization of message with manifest [scala.Some] in [akka.remote.serialization.MiscMessageSerializer]
```

I also note that the [Akka Serialization documentation mentions a different interface](https://doc.akka.io/docs/akka/2.5/serialization.html#programmatic) for using Akka serialization programmatically (`serialization.findSerializerFor`), and that if this interface is used instead, the problem with the `MiscMessageSerializer` is avoided altogether (NB the `manifest = None` argument):

```scala
val system: ActorSystem = ???
val serialization = SerializationExtension(system)
val serializer = serialization.findSerializerFor(classOf[Some[String]])
val serialized: Array[Byte] = serializer.toBinary(Some("thing"))
val deserialized: Some[String] = serializer.fromBinary(serialized, manifest = None).asInstanceOf[Some[String]]
```

## Questions

- Am I configuring Akka serialization as expected, or does play-redis expect `additional-serialization-bindings` to be turned off or something? (Seems that would be a problem for akka-remote.)
- How come not being able to cache an `Option` hasn't been noticed before? Is it that Akka has only enabled the `additional-serialization-bindings` in 2.5 and I'm on a slightly bleeding edge there?
- Is it actually expected to be able to use the `serialization.deserialize[T]` interface to Akka serialization with all the built-in serializers? Is this actually an Akka bug? Or is that interface somewhat "internal" / should the `findSerializerFor` interface be used instead?

## Workaround

Easy enough to wrap my cache entries in a case class, at which point Akka will use Kryo again. So, easy to work around.

**Comment by KarelCemus** [comment]
Too many questions and quite complex issue, I'm not sure I'll be able to cover all your questions. However, I'll try.

First of all, I'm not an expert on Akka nor its serialization, therefore some things might be implemented inefficiently or so, which might result in unexpected behavior. If you encounter any of it, feel free to propose better implementation.

> Am I configuring Akka serialization as expected, or does play-redis expect additional-serialization-bindings to be turned off or something? (Seems that would be a problem for akka-remote.)
>
> How come not being able to cache an Option hasn't been noticed before? Is it that Akka has only enabled the additional-serialization-bindings in 2.5 and I'm on a slightly bleeding edge there?

I answer both these questions at once. Personally, I don't use Akka's additional bindings, I use Kryo for all but primitives (native support in play-redis, cannot be disabled) and Kryo manages all other classes including an `Option`s correctly. Therefore I've not encountered your issue.

> Is it actually expected to be able to use the serialization.deserialize[T] interface to Akka serialization with all the built-in serializers? Is this actually an Akka bug? Or is that interface somewhat "internal" / should the findSerializerFor interface be used instead?

Well, it's never been tested, but it fully supports Akka serializers, therefore there shouldn't exist any reason why it does not work.

See [here](https://github.com/KarelCemus/play-redis/blob/f586c647b23be0d19dba985de0ce8e9b417729d0/src/main/scala/play/api/cache/redis/connector/AkkaSerializer.scala#L70), play-redis calls `serializer.findSerializerFor` to retrieve proper serializer.

But I see that [deserializer](https://github.com/KarelCemus/play-redis/blob/f586c647b23be0d19dba985de0ce8e9b417729d0/src/main/scala/play/api/cache/redis/connector/AkkaSerializer.scala#L122) does not use the same serializer and uses `Serialization` instead. However, if you [look into the `Serialization`](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/serialization/Serialization.scala#L284), you see that it call `serializerFor` internally, therefore there shouldn't be any issue.


Finally, I spent some time debugging your code and I believe it is an issue of the serializer you use. First, play-redis passes serialized string and class of the object into the serializer and expects deserialized object. Then, using the class, Akka properly acquires your `MiscMessageSerializer`, which extends from `SerializerWithStringManifest`. Then, having the class of the data, it translates the class to the manifest [here](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/serialization/Serializer.scala#L150). Unfortunately, it translates the class into the manifest via `c.getName`, which is incorrect. In fact, the manifest should have been `"C"` instead. which results in your issue.


In conclusion, although I am not an expert, I believe that your serializer should be definitelly valid and supported. Unfortunately, considering invalid translation of the class into the manifest inside your serializer, I percieve this as a bug within the Akka and closing the issue.

I hope this helps, unfortunately, I am not able to implement any other workaround.

**Comment by dominics** [comment]
> Unfortunately, considering invalid translation of the class into the manifest inside your serializer, I percieve this as a bug within the Akka and closing the issue.

Opened as #25945

**Comment by dominics** [comment]
@KarelCemus Looks like the `deserialize` method play-redis uses is a legacy API only kept for binary compatibility, and is due to be deprecated (see the upstream issue)

**Comment by KarelCemus** [comment]
I am not sure I am able to use
```
def deserialize(bytes: Array[Byte], serializerId: Int, manifest: String): Try[AnyRef]
```

since not all serializers provide a manifest, I belive. And using class name as a manifest, which is suggested in the discussion, does not work. That is exactly what current implementation does, but it expects `C` instead of class name.

**Comment by dominics** [comment]
Are you able to use [this interface](https://doc.akka.io/docs/akka/2.5/serialization.html#programmatic) which seems to avoid the issue?

**Comment by KarelCemus** [comment]
It seems you are still missing the point. The issue does not lie in the API or play-redis. It lies in the implementation of `MiscMessageSerializer` instead. If I don't pass exact manifest into the serializer, it won't work, since it [does not retrieve a manifest for a class properly](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/serialization/Serializer.scala#L150).

So it is nice that the manifest is optional (literally, when `manifest = None`), but since it cannot resolve it properly, it does not matter and to avoid the issue you must provide the manifest, which play-redis won't as it is not necessary. It provides data and a class, which is enough.

To be even more precise, consider this:

We have  `val foo = Some("value")`, and then [`serializer.manifest(foo)`](https://github.com/akka/akka/blob/master/akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala#L343) is not equal to the manifest used in [`serializer.fromBinary(data,None)`](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/serialization/Serializer.scala#L150)



To resolve your issue you have to patch the `fromBinary` in `MiscMessageSerializer` method as I linked above, since your `MiscMessageSerializer` does not properly inherit from `SerializerWithStringManifest` and introduces a significant inconsistency instead.

**Comment by dominics** [comment]
@KarelCemus I am definitely missing the point. This *works* with `MiscMessageSerializer`, despite you saying it doesn't:

```
val serialized: Array[Byte] = serializer.toBinary(Some("thing"))
val deserialized: Some[String] = serializer.fromBinary(serialized, manifest = None).asInstanceOf[Some[String]]
```

The bit I'm not getting is why, then, you say:

> since it cannot resolve it properly, it does not matter and to avoid the issue you must provide the manifest

Because not providing a manifest avoids the issue nicely. The code you link to with "does not retrieve a manifest for a class properly" only happens if you pass `Some` manifest...

**Comment by KarelCemus** [comment]
Well, I am glad it works when you not provide the manifest, but not all serializer implementing Akka Serializer may safely ignore it. I believe it is better to pass the result class into the serializer when I know it. Or do you disagree?

The issue is, that when you know **the class** (s-ee that type is not a `String`, which is used as a manifest by `MiscMessageSerializer` but `Class[_]`), then it does not resolve properly the string version of the manifest.

This works:
```
val deserialized: Some[String] = serializer.fromBinary(serialized, manifest = None).asInstanceOf[Some[String]]
```

But this does not and that is the problem. And it definetly shoud work as this latter example is even more precise and still correct.
```
val deserialized: Some[String] = serializer.fromBinary(serialized, manifest = Some(classOf[Some[String]]).asInstanceOf[Some[String]]
```

Is it better now?

**Comment by dominics** [comment]
@KarelCemus Just a quick note to say the legacy `deserialize` interface mentioned in this ticket has been marked deprecated in Akka 2.6.0 - see https://github.com/akka/akka/pull/26981

It look like this _won't_ be mentioned in the migration guide for Akka 2.6 (!)

**Comment by KarelCemus** [comment]
Thanks for getting back to this. I'll take a look and hopefully figure out some other approach

**Comment by KarelCemus** [comment]
Opened #215
